tests/std/tests/LWG3018_shared_ptr_function/test.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <memory>
+
+using namespace std;
+
+int val = 0;
+
+struct decrement_val {
+    void operator()(void (*const)(int)) const {
+        --val;
+    }
+};
+
+void add(const int i) {
+    val += i;
+}
+
+int add_return(const int i) {
+    return val += i;
+}
+
+void subtract(const int i) {
+    val -= i;
+}
+
+int main() {
+    {

```

Comment:
I'd like to see a `static_assert` to establish that `shared_ptr<function_type>` isn't constructible from only a pointer-to-`function_type` (i.e., the deleter is necessary).

Also, `noexcept` participates in function types since C++17. I think we need additional coverage guarded by `_HAS_CXX17` for:
* initializing a `shared_ptr`-to-potentially-throwing-function with a pointer to a `noexcept` function
* initializing a `shared_ptr`-to-`noexcept`-function with a pointer to a `noexcept` function
* converting `shared_ptr`-to-`noexcept`-function to `shared_ptr`-to-potentially-throwing-function 
and some `static_assert`s to establish that:
* `shared_ptr<function_type noexcept>` isn't constructible from non-`noexcept` `function_type` even with a deleter
* `shared_ptr`-to-potentially-throwing-function does not convert to `shared_ptr`-to-`noexcept`-function

Reply:
Done